jni/include/knn_extension/faiss/utils/Heap.h

```diff
@@ -0,0 +1,247 @@
+    bh_ids--;
+    size_t i = start_index + 1, i_father;
+
+    while (i > 1) {
+        i_father = i >> 1;
+        if (!C::cmp2(val, bh_val[i_father], group_id, bh_ids[i_father])) {
+            /* the heap structure is ok */
+            break;
+        }
+        bh_val[i] = bh_val[i_father];
+        bh_ids[i] = bh_ids[i_father];
+        (*group_id_to_index)[bh_ids[i]] = i - 1;
+        i = i_father;
+    }
+    bh_val[i] = val;
+    bh_ids[i] = group_id;
+    (*group_id_to_id)[group_id] = id;
+    (*group_id_to_index)[group_id] = i - 1;
+}
+
+/**
+ * From start_index, it compare its value with child node's and swap if needed.
+ * Continue either there is no swap or it reaches the leaf node.
+ *
+ * @param nres        number of search result
+ * @param bh_val      search result, distances from query (updated)
+ * @param bh_ids      search result, ids of vectors (updated)
+ * @param val         distance from query for current vector
+ * @param id          id of current vector
+ * @group_id_to_id    group id to id mapping data (updated)
+ * @group_id_to_index group id to index mapping data (updated)
+ * @group_id          group id of current vector
+ * @start_index       start index to process
+ */
+template <class C>
+static inline void down_heap(
+        int nres,
+        typename C::T* bh_val,
+        typename C::TI* bh_ids,
+        typename C::T val,
+        typename C::TI id,
+        std::unordered_map<typename C::TI, typename C::TI>* group_id_to_id,
+        std::unordered_map<typename C::TI, size_t>* group_id_to_index,
+        typename C::TI group_id,
+        size_t start_index) {
+    bh_val--; /* Use 1-based indexing for easier node->child translation */
+    bh_ids--;
+    size_t i = start_index + 1, i1, i2;
+
+    while (1) {
+        i1 = i << 1;
+        i2 = i1 + 1;
+        if (i1 > nres) {
+            break;
+        }
+
+        // Note that C::cmp2() is a bool function answering
+        // `(a1 > b1) || ((a1 == b1) && (a2 > b2))` for max
+        // heap and same with the `<` sign for min heap.
+        if ((i2 == nres + 1) ||
+            C::cmp2(bh_val[i1], bh_val[i2], bh_ids[i1], bh_ids[i2])) {
+            if (C::cmp2(val, bh_val[i1], group_id, bh_ids[i1])) {
+                break;
+            }
+            bh_val[i] = bh_val[i1];
+            bh_ids[i] = bh_ids[i1];
+            (*group_id_to_index)[bh_ids[i]] = i - 1;
+            i = i1;
+        } else {
+            if (C::cmp2(val, bh_val[i2], group_id, bh_ids[i2])) {
+                break;
+            }
+            bh_val[i] = bh_val[i2];
+            bh_ids[i] = bh_ids[i2];
+            (*group_id_to_index)[bh_ids[i]] = i - 1;
+            i = i2;
+        }
+    }
+    bh_val[i] = val;
+    bh_ids[i] = group_id;
+    (*group_id_to_id)[group_id] = id;
+    (*group_id_to_index)[group_id] = i - 1;
+}
+
+/**
+ * Push the value to the max heap
+ * The group_id should be new which does not exist in bh_ids, group_id_to_id, and group_id_to_index.
```

Comment:
Can you please add how that `group_id` is generated? Also the line itself needs a bit of rewording